File: _designs_001/Add and Search Word - Data structure design/Solution.java
public class Solution {
    public List<Boolean> testWordDictionary(StreamTokenizer tokenizer, List<Double> metrics) throws IOException {
        WordDictionary wd = new WordDictionary();
        List<Boolean> result = new ArrayList<Boolean>();

        int opsCnt = Serializer.deserializeInt(tokenizer);
        for (int i = 0; i < opsCnt; i++) {
            String word;
            double startTime;
            switch (Serializer.deserializeInt(tokenizer)) {
            case 0:
                word = Serializer.deserializeString(tokenizer);
                startTime = System.nanoTime();
                wd.addWord(word);
                metrics.set(0, metrics.get(0) + (double) (System.nanoTime() - startTime) / 1000000L);
                break;
            case 1:
                word = Serializer.deserializeString(tokenizer);
                startTime = System.nanoTime();
                result.add(wd.search(word));
                metrics.set(0, metrics.get(0) + (double) (System.nanoTime() - startTime) / 1000000L);
                break;
            }
        }
        return result;
    }
}

File: _designs_001/Add and Search Word - Data structure design/Solution.rb
class Solution
  def testWordDictionary(inp, metrics)
    metrics[0] = 0
    wd, r = WordDictionary.new, []
    Serializer.deserializeInt(inp).times do
      case Serializer.deserializeInt(inp)
      when 0
        word = Serializer.deserializeString(inp)
        start_time = Time.now
        wd.addWord(word)
        metrics[0] += (Time.now - start_time) * 1000
      when 1
        word = Serializer.deserializeString(inp)
        start_time = Time.now
        r << wd.search(word)
        metrics[0] += (Time.now - start_time) * 1000
      end
    end
    r
  end
end
